Feature/add location feature

itsahmedsalah opened the pull request.

itsahmedsalah committed:
init maps

itsahmedsalah committed:
implemented the location permission in the data and remote

itsahmedsalah committed:
Refactor: Removed LocationWebView component

itsahmedsalah committed:
Merge remote-tracking branch 'origin/develop' into feature/Add-location-Feature

# Conflicts:
#	app/build.gradle.kts
#	app/src/main/java/com/paris_2/san3a/di/dataModule.kt
#	app/src/main/java/com/paris_2/san3a/presentation/San3aScaffold.kt

itsahmedsalah committed:
Integrate domain and viewmodel Koin modules

This commit integrates the `domainModule` and `viewModelModule` into the main `appModule` for dependency injection.

The `domainModule` provides the `GetCurrentLocationUseCase`.

The `Location` entity has been updated to include `countryName`. The `LocationRepositoryImp` has been updated to fetch and provide the `countryName` along with the `cityName`.

itsahmedsalah committed:
Merge remote-tracking branch 'origin/develop' into feature/Add-location-Feature

ahmedNaser7 commented on app/src/main/java/com/paris_2/san3a/data/repository/LocationRepositoryImp.kt:
```diff
@@ -0,0 +1,63 @@
+        val geocoder = Geocoder(context, Locale.ENGLISH)
+        val addresses = geocoder.getFromLocation(latitude, longitude, 1)
+        return if (!addresses.isNullOrEmpty()) {
+            addresses[0].locality ?: addresses[0].subAdminArea ?: "Unknown city"
```
you have to display city name , so i think you have to custom exception for unknow City

ahmedNaser7 commented on app/src/main/java/com/paris_2/san3a/data/repository/LocationRepositoryImp.kt:
```diff
@@ -0,0 +1,63 @@
+        val geocoder = Geocoder(context, Locale.ENGLISH)
+        val addresses = geocoder.getFromLocation(latitude, longitude, 1)
+        return if (!addresses.isNullOrEmpty()) {
+            addresses[0].locality ?: addresses[0].countryName ?: "Unknown city"
```
also here for country

ahmedNaser7 commented on app/src/main/java/com/paris_2/san3a/data/repository/LocationRepositoryImp.kt:
```diff
@@ -0,0 +1,63 @@
+        return fusedLocationProviderClient.getCurrentLocation(
+            currentLocationRequest,
+            null
+        ).await() ?: throw IllegalStateException("No location available")
```
you should handle exception there , why you dont make a domain exception

itsahmedsalah committed:
Implemented exceptions in LocationRepositoryImp

This commit introduces three new exceptions (`UnKnownCityException`, `UnKnownCountryException`, `LocationErrorException`) to handle cases where city, country, or location information is unavailable. These exceptions replace the generic `IllegalStateException` and string literals previously used in `LocationRepositoryImp.kt`.

Additionally, the `osmdroid-android` library has been added as a dependency in `app/build.gradle.kts`.

itsahmedsalah committed:
integrate map screen

This commit integrates the map screen using the OpenStreetMap library.

Key changes:
- Added OpenStreetMap dependencies to `gradle/libs.versions.toml` and `app/build.gradle.kts`.
- Replaced the `WebView` in `LocationWebView.kt` with an `OpenStreetMap` composable.
- The map now displays a marker at a predefined location and allows for camera control.

itsahmedsalah committed:
Merge remote-tracking branch 'origin/develop' into feature/Add-location-Feature

# Conflicts:
#	app/build.gradle.kts
#	app/src/main/java/com/paris_2/san3a/di/appModule.kt
#	app/src/main/java/com/paris_2/san3a/di/dataModule.kt
#	app/src/main/java/com/paris_2/san3a/di/viewModelModule.kt
#	app/src/main/java/com/paris_2/san3a/domain/San3aExceptionTree.kt

itsahmedsalah committed:
feat: Add location pin and location map view

This commit introduces a new location pin icon and implements a location map view component.

The `LocationWebView.kt` file now includes a Composable function that displays an OpenStreetMap view with a marker at the specified latitude and longitude. The map view is configured with specific properties to disable zoom buttons, animations, and gestures.

The `dataModule.kt` file has been updated to include `LocationRepositoryImp` and `LocationRepository` for managing location-related data.

itsahmedsalah committed:
Make the LocationWebView component interactive

The LocationWebView component has been updated to allow users to interact with the map and select a location.

Key changes:
- Added `aLatitude` and `aLongitude` state variables to store the selected latitude and longitude.
- Initialized the marker and camera states with these variables.
- Updated the `onMapClick` listener to update the marker's position and log the selected coordinates.
- Set the initial latitude and longitude to Cairo's coordinates.

itsahmedsalah committed:
Merge remote-tracking branch 'origin/develop' into feature/Add-location-Feature

# Conflicts:
#	app/build.gradle.kts

Mohamed-Elhanafy commented on app/src/main/java/com/paris_2/san3a/di/dataModule.kt:
```diff
@@ -11,12 +12,14 @@ import com.paris_2.san3a.data.source.remote.messages.MessagesRemoteDataSource
     single { AppPreferences(androidApplication().applicationContext) }
     single<FireStoreService> { FireStoreServiceImpl(get()) }
     single { FirebaseFirestore.getInstance() }
+    single<LocationRepository> { LocationRepositoryImp(get()) }
```
use single of

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/data/repository/LocationRepositoryImp.kt:
```diff
@@ -0,0 +1,66 @@
+
+@SuppressLint("MissingPermission")
+class LocationRepositoryImp(private val context: Context) : LocationRepository {
+    private val fusedLocationProviderClient =
+        LocationServices.getFusedLocationProviderClient(context)
```
inject it

Joseph-Sameh-0 replied:
instead of inject the context

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/data/repository/LocationRepositoryImp.kt:
```diff
@@ -0,0 +1,66 @@
+    private fun getCityNameFromCoordinates(latitude: Double, longitude: Double): String {
+        val geocoder = Geocoder(context, Locale.ENGLISH)
+        val addresses = geocoder.getFromLocation(latitude, longitude, 1)
+        return if (!addresses.isNullOrEmpty()) {
```
wrap the all function with try catch or make safe call function for exception handling

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/di/domainModule.kt:
```diff
@@ -0,0 +1,8 @@
+import org.koin.dsl.module
+
+val domainModule = module {
+    single { GetCurrentLocationUseCase(get()) }
```
use single of

itsahmedsalah committed:
Make the LocationWebView component interactive

The LocationWebView component has been updated to allow users to interact with the map and select a location.

itsahmedsalah committed:
Init location mapper and service

This commit introduces the `LocationMapper` to convert `android.location.Location` to `LocationDto` and `LocationServiceImp` to provide the current location using `FusedLocationProviderClient`.
It includes permission checks for location access.

itsahmedsalah committed:
Refactor: Improve location data handling and rename Location entity

This commit refactors the location data handling by:

- Renaming the `Location` entity to `AppLocation` for clarity.
- Introducing a `LocationService` to encapsulate location provider logic.
- Updating `LocationRepositoryImp` to use the new `LocationService`.
- Modifying `LocationMapper` to include a `toLocation()` function for converting `LocationDto` to `AppLocation`.
- Enhancing city and country name retrieval to provide more robust results.

itsahmedsalah committed:
Refactor: Improve location data handling and rename Location entity

This commit refactors the location data handling by:

- Renaming the `Location` entity to `AppLocation` for clarity.
- Introducing a `LocationService` to encapsulate location provider logic.
- Updating `LocationRepositoryImp` to use the new `LocationService`.
- Modifying `LocationMapper` to include a `toLocation()` function for converting `LocationDto` to `AppLocation`.
- Enhancing city and country name retrieval to provide more robust results.

itsahmedsalah committed:
Merge remote-tracking branch 'origin/develop' into feature/Add-location-Feature

# Conflicts:
#	app/src/main/java/com/paris_2/san3a/di/dataModule.kt
